lnrpc/routerrpc/router_backend.go

```diff
@@ -321,3 +340,166 @@ func (r *RouterBackend) UnmarshallRoute(rpcroute *lnrpc.Route) (
 
 	return route, nil
 }
+
+// extractIntentFromSendRequest attempts to parse the SendRequest details
+// required to dispatch a client from the information presented by an RPC
+// client.
+func (r *RouterBackend) extractIntentFromSendRequest(rpcPayReq *PaymentRequest) (
+	*routing.LightningPayment, error) {
+
+	payIntent := &routing.LightningPayment{}
+
+	// Pass along an outgoing channel restriction if specified.
+	if rpcPayReq.OutgoingChanId != 0 {
+		payIntent.OutgoingChannelID = &rpcPayReq.OutgoingChanId
+	}
+
+	// Take cltv limit from request if set.
+	if rpcPayReq.CltvLimit != 0 {
+		cltvLimit := uint32(rpcPayReq.CltvLimit)
+		payIntent.CltvLimit = &cltvLimit
+	}
+
+	// Take fee limit from request.
+	payIntent.FeeLimit = lnwire.NewMSatFromSatoshis(
+		btcutil.Amount(rpcPayReq.FeeLimit),
+	)
+
+	// Set payment attempt timeout.
+	if rpcPayReq.TimeoutSeconds == 0 {
+		return nil, errors.New("timeout_seconds must be specified")
+	}
+
+	payIntent.PayAttemptTimeout = time.Second *
+		time.Duration(rpcPayReq.TimeoutSeconds)
+
+	// If the payment request field isn't blank, then the details of the
+	// invoice are encoded entirely within the encoded payReq.  So we'll
+	// attempt to decode it, populating the payment accordingly.
+	if rpcPayReq.PaymentRequest != "" {
+		switch {
+
+		case len(rpcPayReq.Dest) > 0:
+			return nil, errors.New("dest and payment_request " +
+				"cannot appear together")
+
+		case len(rpcPayReq.PaymentHash) > 0:
+			return nil, errors.New("dest and payment_hash " +
+				"cannot appear together")
+
+		case rpcPayReq.FinalCltvDelta != 0:
+			return nil, errors.New("dest and final_cltv_delta " +
+				"cannot appear together")
+		}
+
+		payReq, err := zpay32.Decode(
+			rpcPayReq.PaymentRequest, r.ActiveNetParams,
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		// Next, we'll ensure that this payreq hasn't already expired.
+		err = validatePayReqExpiry(payReq)
+		if err != nil {
+			return nil, err
+		}
+
+		// If the amount was not included in the invoice, then we let
+		// the payee specify the amount of satoshis they wish to send.
+		// We override the amount to pay with the amount provided from
+		// the payment request.
+		if payReq.MilliSat == nil {
+			if rpcPayReq.Amt == 0 {
+				return nil, errors.New("amount must be " +
+					"specified when paying a zero amount " +
+					"invoice")
+			}
+
+			payIntent.Amount = lnwire.NewMSatFromSatoshis(
+				btcutil.Amount(rpcPayReq.Amt),
+			)
+		} else {
+			if rpcPayReq.Amt != 0 {
+				return nil, errors.New("amount must not be " +
```

Comment:
We can let user overpay if they want, but this seems safer indeed.

Reply:
Yes it seems safer to me like this.